be/src/util/internal_service_recoverable_stub.cpp

```diff
@@ -0,0 +1,164 @@
+// Copyright 2021-present StarRocks, Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "util/internal_service_recoverable_stub.h"
+
+#include "common/config.h"
+
+namespace starrocks {
+
+class RecoverableClosure : public ::google::protobuf::Closure {
+public:
+    RecoverableClosure(std::shared_ptr<starrocks::PInternalService_RecoverableStub> stub,
+                       ::google::protobuf::RpcController* controller, ::google::protobuf::Closure* done)
+            : _stub(stub), _controller(controller), _done(done) {}
+
+    void Run() override {
+        auto* cntl = static_cast<brpc::Controller*>(_controller);
+        if (cntl->Failed() && cntl->ErrorCode() == EHOSTDOWN) {
+            auto st = _stub->reset_channel();
+            if (!st.ok()) {
+                LOG(WARNING) << "Fail to reset channel: " << st.to_string();
+            }
+        }
+        _done->Run();
```

Comment:
Will changing the stub affect _done->Run()?

Reply:
the callback will not use stub